magma/compile_guard.py

```diff
@@ -45,146 +84,51 @@ def __init__(self, name, cond, type):
-            self._rewire_input(port, value)
-            return
-        ref = _get_top_ref(ref)
-        if isinstance(ref, InstRef):
-            # Internal instance driver is okay
-            assert not self._is_external(value)
-            return
-        # TODO(rsetaluri): Do the rest of these.
-        raise NotImplementedError(ref, type(ref))
+    def add_port(self, T: Kind, name: Optional[str] = None) -> Type:
+        if name is None:
+            name = self._new_port_name()
+        return self._add_port(name, T)
 
-    def _process_undriven(self, port):
-        # TODO(rsetaluri): Add other system types
-        if not isinstance(port, Clock):
-            return
-        T = type(port).undirected_t
-        if T in self._system_types_added:
-            return
-        name = str(port.name)
-        self._add_port(name, In(T))
-        self._system_types_added.add(T)
+    def instances(self) -> InstanceCollection:
+        return self._instances.copy()
 
+    def open(self):
+        return self._open()
 
-class _CompileGuard:
-    __index = 0
+    def _new_port_name(self):
+        name = f"port_{self._num_ports}"
+        self._num_ports += 1
+        return name
 
-    def __init__(self, cond: str,
-                 defn_name: Optional[str],
-                 inst_name: Optional[str],
-                 type: Optional[str] = "defined"):
-        self._cond = cond
-        self._defn_name = defn_name
-        self._inst_name = inst_name
-        self._state = None
-        self._type = type
 
-    @staticmethod
-    def _new_name():
-        index = _CompileGuard.__index
-        _CompileGuard.__index += 1
-        return f"CompileGuardCircuit_{index}"
-
-    def __enter__(self):
-        if self._state is not None:
-            raise Exception("Can not enter compile guard multiple times")
-        assert self._state is None
-        if self._defn_name is None:
-            self._defn_name = _CompileGuard._new_name()
-        if self._state is None:
-            ckt = _CompileGuardBuilder(self._defn_name, self._cond, self._type)
-            if self._inst_name is None:
-                ckt.set_instance_name(self._inst_name)
-            ctx_mgr = definition_context_manager(ckt._context)
-            self._state = _CompileGuardState(ckt, ctx_mgr)
-        self._state.ctx_mgr.__enter__()
-
-    def __exit__(self, typ, value, traceback):
-        self._state.ctx_mgr.__exit__(typ, value, traceback)
-        self._state.ckt.pre_finalize()
-
-
-def compile_guard(cond: str,
-                  defn_name: Optional[str] = None,
-                  inst_name: Optional[str] = None,
-                  type: Optional[str] = "defined"):
-    return _CompileGuard(cond, defn_name, inst_name, type)
+def _make_builder(
+        cond: str,
+        defn_name: Optional[str],
+        inst_name: Optional[str],
+        type: str
+) -> _CompileGuardBuilder:
+    if defn_name is None:
+        defn_name = _CompileGuardBuilder.make_default_defn_name()
```

Comment:
Could move this inside the `_CompileGuardBuilder` init logic to make more reuseable